### PR TITLE
fix(ShellEx): Support adding multiple files

### LINKED ISSUE #11711: right click on the Contex menu->GetExt Add files, and nothing happens
### Environment

- Git Extensions 4.2.1.17611
- Build b0c0b284876b63c52c77195ada82a85bfadbb083
- Git 2.42.0.windows.2
- Microsoft Windows NT 10.0.19045.0
- .NET 6.0.25
- DPI 96dpi (no scaling)
- Portable: True
- Microsoft.WindowsDesktop.App Versions

```
    Microsoft.WindowsDesktop.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

### Issue description

I would like to use the context menu to add new files to the git repo, but nothing happens when I use the context menu add file item.

### Steps to reproduce

1, in a git repo, add a new file, for example, a `a.txt` file
2, right click on the file, and press the context menu item "GitExt Add files"
3, nothing happens, no pop up dialog, actually no file is added to the git stage.

### Did this work in previous version of GitExtensions?

Not sure. If I remember correctly, in previous version, I can only add file one by one, I mean if I select multiple files, and click the context menu to add files, only the first file is added.

### Diagnostics

_No response_

### PR DESCRIPTION
**Pull request by mstv**
Fixes #11711

## Proposed changes

- fix(ShellEx): Support adding multiple files
- fix(ShellEx): Write `InstallDir` to registry on registration of the shell extension
- fix(ShellEx): Open GE with current dir as work dir

### Before

N/A

### After

image

## Test methodology 

- manual

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Review by gerhardol — COMMENTED** [review]
+1 have not run or reviewed thoroughly, test succeeded.

File: src/app/GitUI/CommandsDialogs/SettingsDialog/ShellExtension/ShellExtensionManager.cs [hunk 1]
```diff
@@ -42,7 +42,11 @@ public static bool IsRegistered()
         /// <exception cref="FileNotFoundException">If at least one necessary for registration file wasn't found</exception>
         /// <exception cref="Win32Exception">If user canceled elevation dialog (when ex.NativeErrorCode == 1223)</exception>
         /// <exception cref="Exception">Other potential error</exception>
-        public static void Register() => RunRegSvrForShellExtensionDlls("/s {0}");
+        public static void Register()
+        {
+            AppSettings.SetInstallDir(AppSettings.GetGitExtensionsDirectory());
```
Comment (RussKie) [inline src/app/GitUI/CommandsDialogs/SettingsDialog/ShellExtension/ShellExtensionManager.cs:47]: Wouldn't we need the same for other invocation of `RunRegSvrForShellExtensionDlls`?
Wouldn't it make sense to call `AppSettings.SetInstallDir` in `RunRegSvrForShellExtensionDlls`?

File: src/app/GitUI/CommandsDialogs/FormAddFiles.cs [hunk 2]
```diff
@@ -15,13 +15,13 @@ public FormAddFiles(IGitUICommands commands, string? addFile = null)
 
         private void ShowFilesClick(object sender, EventArgs e)
         {
-            string arguments = string.Format("add --dry-run{0} \"{1}\"", force.Checked ? " -f" : "", Filter.Text);
+            string arguments = string.Format("add --dry-run{0} {1}", force.Checked ? " -f" : "", Filter.Text);
```
Comment (RussKie) [inline src/app/GitUI/CommandsDialogs/FormAddFiles.cs:18]: What's the value of `Filter.Text`? Doesn't it need to be quoted? That is `Filter.Text.Quote()`?

File: src/app/GitUI/GitUICommands.cs [hunk 3]
```diff
@@ -1427,7 +1427,7 @@ private bool RunCommandBasedOnArgument(IReadOnlyList<string> args, IReadOnlyDict
                     return true;
                 case "add":
                 case "addfiles":
-                    return StartAddFilesDialog(null, args.Count == 3 ? args[2] : ".");
+                    return StartAddFilesDialog(null, args.Count < 3 ? "." : string.Join(' ', args.Skip(2).Select(file => file.Quote())));
```
Comment (RussKie) [inline src/app/GitUI/GitUICommands.cs:1430]: A comment would be useful here

**Comment by RussKie** [comment]
Noting that the fix has been confirmed by @asmwarrior in https://github.com/gitextensions/gitextensions/issues/11711#issuecomment-2143792185

File: src/app/GitUI/CommandsDialogs/SettingsDialog/ShellExtension/ShellExtensionManager.cs (hunk 1 above)
Comment (mstv) [reply src/app/GitUI/CommandsDialogs/SettingsDialog/ShellExtension/ShellExtensionManager.cs:47]: > Wouldn't it make sense to call AppSettings.SetInstallDir in RunRegSvrForShellExtensionDlls?

Not for deregistration.

> Wouldn't we need the same for other invocation of RunRegSvrForShellExtensionDlls?

Deregistration is the only other caller.

File: src/app/GitUI/CommandsDialogs/FormAddFiles.cs (hunk 2 above)
Comment (mstv) [reply src/app/GitUI/CommandsDialogs/FormAddFiles.cs:18]: > What's the value of Filter.Text? Doesn't it need to be quoted? That is Filter.Text.Quote()?

Command line arguments have already been quoted (refer to the screenshot in the PR description). Other callsites do not fill the textbox.
So if the user removes quotes from file names with spaces, it is their fault.

Comment (RussKie) [reply src/app/GitUI/CommandsDialogs/FormAddFiles.cs:18]: Makes sense

File: src/app/GitUI/CommandsDialogs/SettingsDialog/ShellExtension/ShellExtensionManager.cs (hunk 1 above)
Comment (RussKie) [reply src/app/GitUI/CommandsDialogs/SettingsDialog/ShellExtension/ShellExtensionManager.cs:47]: Makes sense

**Review by RussKie — APPROVED** [review]

**Pull request merged**